# Synaptic-Proteome-Diversity
This repository includes the code develop for the scientifc article entiteled 'Synaptic proteome diversity is primarily driven by gene regulation of glutamate receptors and their regulatory proteins'.

Article Abstract,

Electrophysiological features of excitatory synapses vary widely throughout the brain, granting neuronal circuits the ability to decode and store diverse patterns of information. Synapses formed by the same neurons have similar electrophysiological characteristics, belonging to the same type. However, these are generally confined to microscopic brain regions, precluding their proteomic analysis. This has greatly limited our ability to investigate the molecular basis of synaptic physiology. Here we introduce a procedure to characterise the proteome of individual synaptic types. We reveal a remarkable proteomic diversity among the synaptic types of the trisynaptic circuit. Differentially expressed proteins participate in well-known synaptic processes, controlling the signalling pathways preferentially used among diverse synapses. Noteworthy, all synaptic types differentially express proteins directly involved in the function of glutamate receptors. Moreover, neuron-specific gene expression programs would participate in their regulation. Indeed, genes coding for these proteins exhibit such distinct expression profiles between neuronal types that they greatly contribute to their classification. Our data is an important resource for exploring the molecular mechanisms behind electrophysiological properties of different hippocampal synaptic types. Our combined analysis of proteomics and transcriptomics data uncovers a previously unrecognised neuron-specific transcriptomic control of synaptic proteome diversity, directed towards the regulation of glutamate receptors and their regulatory proteins.

Authors and Affiliations,

Rita Reig-Viader1,2*, Diego del Castillo-Berges1,2*, Albert Burgas-Pau1,2,3,4*, Daniel Arco-Alonso1*, David Ramos-Vicente1,5, Carlos Sindreu6, Àlex Bayés1,2#

1.	Molecular Physiology of the Synapse Laboratory, Institut de Recerca Sant Pau (IR Sant Pau), Barcelona, Spain.
2.	Universitat Autònoma de Barcelona (UAB), Bellaterra (Cerdanyola del Vallès), Spain.
3.	Unitat mixta d’Investigació IRTA-UAB en Sanitat Animal. Centre de Recerca en Sanitat Animal (CReSA), Bellaterra (Cerdanyola del Vallès), Spain.  
4.	IRTA. Programa de Sanitat Animal. Centre de Recerca en Sanitat Animal (CreSA), Bellaterra, 08193, Catalonia. Spain.
5.	Neurodegenerative Diseases Research Group, Vall d’Hebron Research Institute, Centre for Networked Biomedical Research on Neurodegenerative Diseases (CIBERNED), Barcelona, Catalonia, Spain.
6.	Institut d'Investigacions Biomèdiques August Pi i Sunyer (IDIBAPS), Barcelona, Spain.

* These authors contributed equally to the work 
# Corresponding author:

Àlex Bayés, Molecular Physiology of the Synapse Laboratory, Institut de Recerca Sant Pau (IR Sant Pau), Barcelona, Spain., C/Sant Quintí, 77-79, 08041 Barcelona, Spain.
Email: abayesp@santpau.cat 
